osbuild/sources.py

```diff
@@ -1,14 +1,76 @@
+import concurrent.futures
 import abc
 import contextlib
 import os
 import json
 import tempfile
+from typing import List, Dict
+from enum import Enum
 
+from osbuild.formats_common import OSBuildError
 from . import host
 from .objectstore import ObjectStore
 from .util.types import PathLike
 
 
+class SourceErrorKind(Enum):
+    UNKNOWN = 0
+    CHECKSUM = 1
+    # OSTree errors
+    OSTREE_INIT = 2
+    OSTREE_CONFIG = 3
+    OSTREE_REMOTE_ADD = 4
+    OSTREE_GPG_IMPORT = 5
+    OSTREE_PULL = 6
+    OSTREE_REMOTE_DELETE = 7
+    # Skopeo errors
+    SKOPEO_COPY = 8
+    SKOPEO_INSPECT = 9
+    # Curl errors
+    CURL_BAD_INPUT = 10
+    CURL_INTERNAL_ERROR = 11
+    CURL_NETWORKING = 12
+    CURL_CREDENTIALS = 13
+    CURL_STORAGE = 14
+    CURL_SECURITY = 15
+
+
+class SourceError(OSBuildError):
+    _ID = 1000
+
+    def __init__(self, kind: SourceErrorKind, message: str, source: str):
+        self.kind = kind
+        self.message = message
+        self.source = source
+        super().__init__(message)
+
+    @property
+    def code(self):
+        return SourceErrorKind(self.kind).name
+
+    @property
+    def details(self):
+        return {"source": self.source}
```

Comment:
This is something that is is format specific and should not be on the error class.